lib/auth/keystore/keystore.go

```diff
@@ -37,17 +39,86 @@ type KeyStore interface {
 
-	// GetSSHSigner selects the local SSH keypair and returns an ssh.Signer.
+	// GetSSHSigner selects the local SSH keypair from the CA ActiveKeys and
+	// returns an ssh.Signer.
 	GetSSHSigner(ca types.CertAuthority) (ssh.Signer, error)
 
-	// GetJWTSigner selects the local JWT keypair and returns a crypto.Signer
+	// GetJWTSigner selects the local JWT keypair from the CA ActiveKeys and
+	// returns a crypto.Signer.
 	GetJWTSigner(ca types.CertAuthority) (crypto.Signer, error)
 
-	// DeleteKey deletes the given key from the KeyStore
+	// NewSSHKeyPair creates and returns a new SSHKeyPair.
+	NewSSHKeyPair() (*types.SSHKeyPair, error)
+
+	// NewTLSKeyPair creates and returns a new TLSKeyPair.
+	NewTLSKeyPair(clusterName string) (*types.TLSKeyPair, error)
+
+	// NewJWTKeyPair creates and returns a new JWTKeyPair.
+	NewJWTKeyPair() (*types.JWTKeyPair, error)
+
+	// HasLocalActiveKeys returns true if the given CA has any active keys that
+	// are usable with this KeyStore.
+	HasLocalActiveKeys(ca types.CertAuthority) bool
+
+	// HasLocalActiveKeys returns true if the given CA has any additional
+	// trusted keys that are usable with this KeyStore.
+	HasLocalAdditionalKeys(ca types.CertAuthority) bool
+
+	// DeleteKey deletes the given key from the KeyStore.
 	DeleteKey(keyID []byte) error
+
+	// DeleteUnusedKeys deletes all keys from the KeyStore if they are:
+	// 1. Labeled by this KeyStore when they were created
+	// 2. Not included in the argument usedKeys
+	DeleteUnusedKeys(usedKeys [][]byte) error
+
+	// GetAdditionalTrustedSSHSigner selects the local SSH keypair from the CA
+	// AdditionalTrustedKeys and returns an ssh.Signer.
+	GetAdditionalTrustedSSHSigner(ca types.CertAuthority) (ssh.Signer, error)
+
+	// GetAdditionalTrustedTLSCertAndSigner selects the local TLS keypair from
+	// the CA AdditionalTrustedKeys and returns the PEM-encoded TLS cert and a
+	// crypto.Signer.
+	GetAdditionalTrustedTLSCertAndSigner(ca types.CertAuthority) ([]byte, crypto.Signer, error)
+}
+
+// Config is used to pass KeyStore configuration to NewKeyStore.
+type Config struct {
+	// RSAKeyPairSource is a function which returns raw keypairs to be used if
+	// an hsm is not configured
+	RSAKeyPairSource RSAKeyPairSource
+
+	// Path is the path to the PKCS11 module.
+	Path string
+	// SlotNumber points to the PKCS11 slot to use, or nil if slot is unused.
+	SlotNumber *int
+	// TokenLabel is the label of the PKCS11 token to use.
+	TokenLabel string
+	// Pin is the PKCS11 pin for the given token.
+	Pin string
+	// HostUUID is the UUID of the local auth server this HSM is connected to.
+	HostUUID string
+}
+
+// NewKeyStore returns a new KeyStore
+func NewKeyStore(cfg Config) (KeyStore, error) {
+	if cfg.Path == "" {
+		if cfg.RSAKeyPairSource == nil {
+			return nil, trace.BadParameter("must provide one of Path or RSAKeyPairSource in keystore.Config")
+		}
+		return NewRawKeyStore(&RawConfig{cfg.RSAKeyPairSource}), nil
+	}
```

Comment:
add a `CheckAndSetDefaults` method on the `Config` and make sure all other required fields are set (e.g. `HostUUID`)

Reply:
done